src/config/application.rs

```diff
@@ -1,13 +1,63 @@
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use regex::Regex;
 use serde::{Deserialize, Deserializer};
 
 // TODO: Use trait to allow only either `only` or `not`
 #[derive(Debug, Deserialize)]
 #[serde(deny_unknown_fields)]
 pub struct Application {
-    #[serde(default, deserialize_with = "deserialize_string_or_vec")]
-    pub only: Option<Vec<String>>,
-    #[serde(default, deserialize_with = "deserialize_string_or_vec")]
-    pub not: Option<Vec<String>>,
+    #[serde(default, deserialize_with = "deserialize_matchers")]
+    pub only: Option<Vec<ApplicationMatcher>>,
+    #[serde(default, deserialize_with = "deserialize_matchers")]
+    pub not: Option<Vec<ApplicationMatcher>>,
+}
+
+#[derive(Debug)]
+pub enum ApplicationMatcher {
+    Literal(String),
+    Regex(Regex),
+}
+
+impl ApplicationMatcher {
+    pub fn matches(&self, name: &str) -> bool {
+        match &self {
+            ApplicationMatcher::Literal(s) => s == name,
+            ApplicationMatcher::Regex(r) => r.is_match(name),
+        }
+    }
+}
+
+impl FromStr for ApplicationMatcher {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let first_char = s.chars().next();
+        match first_char {
+            None => Err(anyhow!("Empty application name")),
+            Some('~') if s.len() == 1 => Err(anyhow!("Empty regular expression for application name")),
+            Some('~') => Ok(ApplicationMatcher::Regex(Regex::new(&s[1..])?)),
```

Comment:
I'm not used to the syntax of declaring a regular expression by starting it with `~`. How about wrapping a regex with `/`s [like CircleCI](https://circleci.com/docs/workflows#using-regular-expressions-to-filter-tags-and-branches), e.g. `/regex/`?

Also, it'd be nice to test parsing it at `src/config/tests.rs`.

Reply:
Thanks for the review! Just two quick questions before I move forward:

1. Is there a straightforward way to unescape PCRE regular expressions and create `regex::Regex`? I'm happy to write a simple unescape function but just want to confirm I'm not missing anything
2. Do you also want it to support modifiers, e.g. `/^minecraft/i` for case-insensitive matching?

Reply:
> Is there a straightforward way to unescape PCRE regular expressions and create regex::Regex? I'm happy to write a simple unescape function but just want to confirm I'm not missing anything

I don't know of one. I was just thinking about checking if it starts and ends with a `/` and removing them.

> Do you also want it to support modifiers, e.g. `/^minecraft/i` for case-insensitive matching?

No, I don't. For simplicity, I prefer starting from supporting no flags. I don't think the application name would fluctuate like minecraft, MineCraft, or Minecraft either.

Thanks for asking :slightly_smiling_face:

Reply:
Pushed. Could you have a look again?